### ISSUE TITLE
bodyOption with a JSON output type 400's for an empty body

### ISSUE THREAD

**Issue by JoeEnnever** [finagle/finch#795]
We ran into this when upgrading from 0.12.0 to 0.15.0. Running this code:

```scala
import io.finch._
import io.finch.circe._
import io.circe._
import io.circe.generic.auto._
import com.twitter.finagle.Http
import com.twitter.util.Await

case class Foo(foo: Option[Int] = None)

object BodyOption {
  def bodyOptEndpoint: Endpoint[String] = post("test" :: bodyOption[Foo, Application.Json]).mapOutput(_ => Ok("success"))

  def main(args: Array[String]): Unit = {
    Await.result(Http.server.serve(":8080", bodyOptEndpoint.toService))
  }
}
```
And hitting the endpoint:

```bash
$ curl -i -X POST localhost:8080/test
HTTP/1.1 400 Bad Request
Date: Tue, 13 Jun 2017 19:48:16 GMT
Content-Length: 0
```

In 0.12.0, `bodyOption` would have `None` as the output.

**Comment by vkostyukov** [comment]
Thanks @JoeEnnever! This is a bug in our body implementation that was surfaced by switching the transport over to Netty 4. As a workaround, you can stick to Netty 3 for now  (see release notes on how to do that).

Here is what happens. Netty 4 (when streaming is disabled) [sets `Content-Length` header to zero](https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectAggregator.java#L215-L227) for each incoming message (even it wasn't set before). Netty 3 didn't do that. By some reason, we treat 0-length bodies as non-empty so we try to parse it and fail.

I will fix it and backport the fix to 0.15 (except the 0.15.1 release tonight).

**Comment by vkostyukov** [comment]
The fix was published as [0.15.1](https://github.com/finagle/finch/releases/tag/0.15.1) patch release. Thanks again, @JoeEnnever for reporting this!
